Examples/ByFeature/FunctionalInjection/ShoppingSteps.fs

```diff
@@ -1,17 +1,36 @@
 ﻿module ShoppingSteps
 
+open System
 open NUnit.Framework
 open TickSpec
 
-type Wallet = WalletDollars of int
+type CatalogItem = {Id: int; Description: string; Price: decimal}
+type Year = Year of int
+type Catalog = Catalog of Year:Year * Items:(CatalogItem [])
+type Catalogs = Map<Year, Catalog>
+type Receipt = Receipt of Date:DateTime * Lines:(string [])
 
-let [<Given>] ``I have (.*) dollars in my wallet`` (amount:int) =
-    WalletDollars amount // The output of function gets stored by type
+let [<Given>] ``no catalogs prior to 2018``() =
+    Map.empty: Catalogs
 
-// Extra parameters are restored from store by its type, in this case it is the wallet
-let [<When>] ``I buy an item for (.*) dollars`` cost (WalletDollars wallet) =
-    // Wallet value is replaced in store thanks to this output
-    WalletDollars (wallet - cost)
+// Table parameters can be bound to record type arrays
+let [<Given>] ``the (.+) product catalog:`` (yearInt: int) (items: CatalogItem[]) (catalogs: Catalogs) =
+    let year = Year yearInt
+    let catalog = Catalog(year, items)
+    Map.add year catalog catalogs
 
-let [<Then>] ``My wallet contains (.*) dollars`` (amount:int) (WalletDollars wallet) =
-    Assert.AreEqual(amount, wallet)
\ No newline at end of file
+// Bullet list is bound to string[]
+let [<When>] ``I make a purchase on (.*):`` (purchaseDate: DateTime) (orderItems: string[]) (catalogs: Catalogs) =
+    let Catalog(Items=catalogItems) = catalogs |> Map.find (Year purchaseDate.Year)
+    let receiptLines =
+        orderItems
+        |> Array.map (fun x -> catalogItems |> Array.find (fun y -> y.Description = x))
```

Comment:
Chained Array.map calls are generally frowned upon and/or can distract some readers. Consider defining a small helper function and then passing that to a single Array.map. Alternately use an array comprehension :- `[| fox x in orderItems -> let i = .... ; .... |]`